### ISSUE TITLE
Sails 0.11.0-rc10 config.sockets.beforeConnect is never called

### ISSUE THREAD

**Issue by tavai89** [balderdashy/sails#2613]
Hi Guys, I just updated to the sails@0.11.0-rc10 and did the sockets migration steps from the migration documentation. I noticed a big problem. beforeConnect function which should used now instead of the authorization function is never called.

Do anybody have the same problem?

Thanks,

Taras

**Comment by wmoore** [comment]
Same behavior here - the old onConnect event is called (and displays a depreciation warning), but the beforeConnect event is never called as of sails@0.11.0-rc10, I saw this issue in previous 0.11-rc* as well.

Walter

**Comment by dadleyy** [comment]
I'm also having this problem, here is my `config/sockets.js` file:

```
module.exports.sockets = {

  path: '/sock',

  beforeConnect: function(handshake, cb) {
    sails.log(arguments);
    return cb(null, true);
  },

  grant3rdPartyCookie: true,

  origins: '*:*'

};
```

and I open a socket from the browser-side application using (coffee):

```
io.sails.autoConnect = false
base_socket = io.socket

loaded_config = (data) ->
  base_socket.url = data.urls.api
  base_socket.path = '/sock'
  base_socket._connect()
```

**Comment by Dallas62** [comment]
From https://github.com/balderdashy/sails/blob/master/0.11-migration-guide.md



####  `onConnect` lifecycle callback

> **tldr;**
>
> Remove your `onConnect` function from `config/sockets.js`.

The `onConnect` lifecycle callback has been deprecated.  Instead, if you need to do something when a new socket is connected, send a request from the newly-connected client to do so.  The purpose of `onConnect` was always for optimizing performance (eliminating the need to do this initial extra round-trip with the server), yet its use can lead to confusion and race conditions. If you desperately need to eliminate the server roundtrip, you can bind a handler directly on `sails.io.on('connect', function (newlyConnectedSocket){})` in your bootstrap function (`config/bootstrap.js`). However, note that this is discouraged.  Unless you're facing _true_ production performance issues, you should use the strategy mentioned above for your "on connection" logic (i.e. send an initial request from the client after the socket connects).  Socket requests are lightweight, so this doesn't add any tangible overhead to your application, and it will help make your code more predictable.


`onConnect` no longer exist.

Use a controller to do the same work of `onConnect`

**Comment by sgress454** [comment]
Thanks for the heads-up, the tests are passing but maybe there's something we failed to put in the migration guide.... cc: @mikermcneil

**Comment by cpttripzz** [comment]
+1 pretty big bug

**Comment by imsnif** [comment]
We're currently solving this (somewhat unesthetically) by doing this in bootstrap.js:

  sails.io.on("connect",function(newSocket){
    //Some logic as to whether we accept the connection or not based on newSocket
    newSocket.disconnect(); //Or not...
  });

So yes - a fix would be awesome.

**Comment by tobalsgithub** [comment]
+1

**Comment by shaunpersad** [comment]
+1

**Comment by tobalsgithub** [comment]
Is this an issue with socket.io itself? I'm doing a little debugging, and see that the beforeConnect function does get passed to socket.io appropriately, but when socket.io goes to attach the server, it's overwriting the option.

https://github.com/Automattic/socket.io/blob/master/lib/index.js#L198-L241

I'm specifically looking at line 225. `opts.allowRequest` is correct before this line. Seems like maybe it should be
```opts.allowRequest = opts.allowRequest || this.checkRequest.bind(this);```

When I add this in, the beforeConnect function gets called just fine.

**Comment by tobalsgithub** [comment]
Here's the commit that made the change to how opts.allowRequest is set. I'll open an issue with the socket.io folks.

https://github.com/Automattic/socket.io/commit/463d7a16a139981f1eead005b388f720be0989bd#diff-6d186b954a58d5bb740f73d84fe39073

**Comment by tobalsgithub** [comment]
Issue on the socket.io project: https://github.com/Automattic/socket.io/issues/2050

**Comment by tobalsgithub** [comment]
Pull request to add this ability back in, if the socket.io guys will take it.

https://github.com/Automattic/socket.io/pull/2051

**Comment by TomKaltz** [comment]
+1

**Comment by woss** [comment]
+1

**Comment by mahasen** [comment]
:+1: @tobalsgithub

**Comment by leedm777** [comment]
@tobalsgithub It's been several months with no comments on either your issue or PR into Socket.io. Is there any way to workaround this issue so we can authentication Socket.io connections?

**Comment by tobalsgithub** [comment]
@leedm777 The best I've seen is what @imsnif suggested above. In `bootstrap.js`, use the `connection` event to initiate an authentication process. 
```
sails.io.on('connection', function(newSocket){
  // check if connection should be allowed
  // if not...
  newSocket.disconnect();
  // if so, continue with whatever
});
``` 

Not super elegant, and can be weird to grab whatever credentials are being used for authentication, but we were able to get it to work for us. 

Not seeing much movement at all on the socket.io side. The last few months have been really quiet with just a few `package.json` updates.

**Comment by sgress454** [comment]
Hey guys, thanks for keeping up with this.  Closing on our end since it definitely looks to be a socket.io issue, and unfortunately not something we can fix inside of Sails or sails-hook-sockets.  We'll keep an eye on @tobalsgithub's PR and issue w/ Socket.io, hopefully @rauchg can get to it soon.

**Comment by luislobo** [comment]
@sgress454 the patch has been merged yesterday and we must keep an eye as a new version might appear soon with the fix

**Comment by crobinson42** [comment]
still at socket v1.3.7? What's the next version we should be looking for?
